### ISSUE TITLE
Add fuzz testing

### ISSUE THREAD

**Issue by elsirion** [fedimint/fedimint#287]
Add fuzz testing for:

* Parsing logic receiving untrusted input
* Network stack
* maybe some consensus logic/modules

[See rust-bitcoin for examples.](https://github.com/rust-bitcoin/rust-bitcoin/tree/master/fuzz/fuzz_targets)

**Comment by NicolaLS** [comment]
After doing some research I think that we can use [hongfuzz](https://github.com/rust-fuzz/honggfuzz-rs) this is also the fuzzer rust-bitcoin uses in their CI (they also have `afl` and `cargo-fuzz` available but I'd just go with `hong_fuzz`
Fuzzing e. g. `client-lib utils functions` will be easy, but I wanted to start fuzzing the API endpoints of the `fedimintd` server.
I haven't really found a good approach of doing this..My last thought was to expose the handlers (the closures) in some way (maybe in a `HashMap<'static str, FnOnce>` or something like that (It feels like a dumb take while writing this (also that is not correct syntax I think) but the idea comes across. Any ideas @elsirion @justinmoon @jkitman ?

**Comment by justinmoon** [comment]
I don't know anything about fuzzing. @m1sterc001guy any thoughts here?

**Comment by m1sterc001guy** [comment]
Haven't had time to sufficiently look into this to have an opinion. Why can't the fedimintd endpoint be directly fuzzed instead of the handlers? @NicolaLS let me know if there's something I can help test.

**Comment by wilfredallyn** [comment]
I saw a blogpost about [fuzzing in BDK](https://afilini.com/blog/fuzz-your-mother-if-you-want-fuzz/) that could be helpful.

I also saw that [LDK](https://github.com/lightningdevkit/rust-lightning/tree/main/fuzz) had a helpful readme about their fuzzing

**Comment by elsirion** [comment]
> Why can't the fedimintd endpoint be directly fuzzed instead of the handlers?

Wouldn't that increase latency while possibly reducing the capability of the fuzzer to learn because now there's all that HTTP stuff in between it and the thing we want to test?

**Comment by NicolaLS** [comment]
> I saw a blogpost about [fuzzing in BDK](https://afilini.com/blog/fuzz-your-mother-if-you-want-fuzz/) that could be helpful.
> 
> I also saw that [LDK](https://github.com/lightningdevkit/rust-lightning/tree/main/fuzz) had a helpful readme about their fuzzing

ty, good information

**Comment by NicolaLS** [comment]
@m1sterc001guy my problem was running the server in the fuzzer, doing requests with the arbitrary fuzz data should be as straight forward as fuzzing some function. I think this _could_ be done by stealing some setup from the integration tests we have right now but we want to mock all networking in them anyway because of #493 (also addressing @elsirion concerns)
TLDR: Maybe we should fuzz the endpoints after #493 (in a mocked way) and just start fuzzing e.g. utility functions now

**Comment by justinmoon** [comment]
> I saw a blogpost about [fuzzing in BDK](https://afilini.com/blog/fuzz-your-mother-if-you-want-fuzz/) that could be helpful.
> 
> I also saw that [LDK](https://github.com/lightningdevkit/rust-lightning/tree/main/fuzz) had a helpful readme about their fuzzing

These are really nice posts.

**Comment by dpc** [comment]
Largely done.
